### PR TITLE
[inductor] add cat_linear_fused pre-grad pass

### PR DESCRIPTION
**Pull request by reger-men**
Motivation

`F.linear(torch.cat([t_0, ...], dim=-1), W, b)` materialises the cat on the forward path (and its grad on the backward path). On bf16 GEMMs that extra round-trip through HBM is a measurable bandwidth cost.

What changed

New pre-grad FX pass that rewrites `F.linear(cat([t_0, ...], dim=-1), W, b)` into a reduce-sum of per-piece `F.linear` calls on contiguous weight slices, so the cat never has to be materialised. Conservative gating: fires only when every cat operand is a last-dim slice with the same leading shape, bias is None (or only on the first partial linear), total cat width is below a cap, and the cat axis is the last axis (negative indexing handled). Off by default; opt in via `torch._inductor.config.pre_grad_custom_pass`. The new file changes no default code path.

Test plan

- matcher unit tests: canonical fires, 3-part fires, rejects more than MAX_PARTS, rejects a piece below MIN_PIECE_WIDTH, rejects total above MAX_TOTAL_CAT_WIDTH, rejects a mul-parented operand, rejects a cat on a non-last axis
- integration: `torch.compile` with the pass installed matches eager within 1e-4 and bumps the `cat_linear_fused` counter; with the pass not installed the counter stays 0

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/fx_passes/cat_linear_fused.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+def fuse_cat_linear_in_graph(graph: fx.Graph) -> int:
+    """Run the matcher across `graph`, returning the number of rewrites."""
+    n = 0
+    linear_target = torch.nn.functional.linear
+    add_target = operator.add
+
+    candidates = [
+        node
+        for node in list(graph.nodes)
+        if node.op == "call_function" and _is_linear(node)
```
Comment (KareemMusleh) [inline torch/_inductor/fx_passes/cat_linear_fused.py]: nit: _is_linear already checks that `node.op == "call_function"`

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Review by ezyang — CHANGES_REQUESTED** [review]
There's not a good reason for this to live in core if it's not hooked up to anything. I don't know how to assess the risk of turning this optimization on by default

**Comment by reger-men** [comment]
Reworked this into the group_batch_fusion registry as an opt-in fusion, superseded by #187074. Closing this one.
